klaytn/klaytn
# Fixed corrupt governance state after rewinding chain

## Proposed changes

Added additional clear logics after  `debug.SetHead`.

### Problem Statement
The existing `setHead` implementation considered only for the block data such as receipts and transactions. This PR addresses governance (header and contract) and staking state.

### What PR Does
```
---------------------
setHead (remove transaction, receipt, and block)
remove governance database
---------------------
Post-task
  1. Initiailize snapshot cache
  2. Initiailize governance cache and last state block number
```

Compared to two related PRs (https://github.com/klaytn/klaytn/pull/1186, https://github.com/klaytn/klaytn/pull/1756), this PR does not bridge additional implementation to fix this problem. I just defined the post-task and these utilize the existing implementation.

### Cases
Case 1: header governance (+ cache)
  - Verification error fixed
  - Delete local database and memory cache

Case 2: contract governance
  - We don't need to put additional workaround because the contract-based governance state does not have a dependency on local MiscDB and in-memory cache.

Case 3: staking info DB (+ cache)
  - Delete local database and cache

Case 4: Other DBs (i.e., state DB, MiscDB, etc)
  - There seems not a strong motivation to remove existing state DB by the `setHead` call.

### Bug fix

The original codebase contained a bug where a discrepancy arises between the blockchain state and the headerchain state when lost state data appears because blockchain goes further delete operation exceeding the origin target number which is the origin block number to be rewound. Fixed in this PR together.

### Test
[Prepration]
```
epoch = 10
# of validators = 4
```

[Execution]
1. Vote the following item at the beginning of the network. `governance.vote("reward.mintingamount","123")`
2. `klay.getHeader(5)` contains non-empty vote data
3. `klay.getHeader(10)` contains non-empty governance data
4. Let's rewind 1,2,3,4,5,6,7,8,9,10
    ex) `debug.setHead(7)`
5. Wait to be synced from connected peers
    - Before this PR: A bad block appears at the time of verifying block number 10, which contains non-empty governance data.
    - After this PR: Fixed
6. Catch up the latest block

[Secanrio test]
(Preparation is the same with above)

1. Vote at block 33
2. Terminate a node at block 36
3. Restart the node and wait to sync latest block number
4. Failed to verify block 40 that contains non-empty governnace data
5. call `klay.setHead(30)`.
6. Successfully catchup the latest block number
7. (additional) call `klay.setHead()` with `10, 20, 30, 33, 35, 40, n` works.

## Types of changes

Please put an x in the boxes related to your change.


## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
